### PR TITLE
Proposing upload() to handle multipart request

### PR DESCRIPTION
**Pull request by dstpierre**
Continuing PR #370 

This PR handles the changes to `open()` to support binary format and a way to post multipart requests via a proposed `upload()` function.

It could be done via the `post()` function, but that would involved either looking at all parameters to detect if there's at least one `FileData` and in that case use the multipart format. Another option would be to check for the presence of `multipart/form-data` in the headers.

I think that since `goja` is using heavy reflection, solution A might add some overhead to http requests (I do not have benchmarked or anything).

And the 2nd one would required the users to remember to set the `content-type: "multipart/form-data"`.

I'm posting the PR even though it is not ready so I can get earlier reviews.

### CODE REVIEW

File: js/initcontext_test.go [hunk 1]
```diff
@@ -281,3 +284,83 @@ func TestInitContextOpen(t *testing.T) {
...
+				return true;
+			}
+			`, svr.URL)),
+	}, fs)
+	assert.NoError(t, err)
+
+	bi, err := b.Instantiate()
+	assert.NoError(t, err)
+
+	v, err := bi.Default(goja.Undefined())
```
Comment (dstpierre) [inline js/initcontext_test.go]: @liclac this is the line I was talking about on Slack

**Comment by robingustafsson** [comment]
I think having an `http.upload()` API is too limiting in terms of what you can do with multipart requests. I'd like to propose a more versatile API along the lines of:

```js
import mime from "k6/mime";

let myFile = open("photo.jpg", "b");

export default function() {
    let mp = new mime.MultipartMessage("---optional-custom-boundary")  // If boundary is not specified one would be auto-generated
    mp.addPart("file", myFile, { filename: "override-filename.jpg", headers: { "Content-Type": "application/octet-stream" } });
    mp.addPart("comment", "This is a great photo!")

    // Send a multipart body using any of the supported request methods
    http.post("https://example.com/upload", mp);
    http.put("https://example.com/upload", mp);
    http.patch("https://example.com/upload", mp);

    // ...even batch API
    http.batch([
        [ "POST", "https://example.com/upload", mp ],
        [ "PUT", "https://example.com/upload", mp ],
        [ "PATCH", "https://example.com/upload", mp ]
    ]);
}
```

This way you can:
- Use multipart bodies with all supported `http` module request functions (including `http.batch`).
- Create multipart requests with multiple parts, not just one part/file/field
- Set custom per-part headers

Or, as you said @dstpierre, we could extend the handling of a JS object being passed as the request body and if we detect at least one file field in the JS object or an explicit `multipart/*` header we create a multipart request body. This API could also be made to handle custom per-part headers by accepting JS object values like `{ "value": myFile, "headers": { "Content-Type": "application/octet-stream" } }`.

Thoughts?

**Comment by dstpierre** [comment]
@robingustafsson I like the flexibility exposing a `mime` module would bring. But it would be quicker to have the `post`, `put`, and `patch` functions to handle file upload.

A `mime` module could be added later if there's any use case. The amount of documentation will be minimal changing the existing functions vs. adding an entire module.

```js
import http from "k6/http";
let binFile = open("/path/to/file.bin", "b");

export default function() {
  let data = {
    field: "this is a standard form field",
    file: binFile
  };
  let res = http.post("url_here", data);
}
```

Does that sounds like a good solution?

**Comment by liclac** [comment]
1. So as I commented on #370, binary data should be either `[]byte`s or `ArrayBuffer`-like (which we can scrap once I PR an `ArrayBuffer` JS-side interface for `[]byte`s to goja, whenever that happens since it'll be on my own time).

   The interface idea was something I blurted out as a possible solution without much thought, that was taken, slightly misinterpreted, and ran with.

2. This is how I'd probably do it; similar to @dstpierre's solution:

   ```js
   import http from "k6/http";
   import { b64encode } from "k6/encoding";

   let myFile = open("image.png", "b");

   export default function() {
     http.post("http://example.com/", {
       "field": "standard form field",
       "file": http.file(myFile, "image.png"), // filename is optional
       "file2": http.file(b64encode(myFile), null, {
         "Content-Transfer-Encoding": "base64",
       }),
     });
   }
   ```

**Comment by dstpierre** [comment]
The last commit should be more aligned with the desired usage.

@liclac I cannot seems to find why the `TestRequestWithBinaryFile` test in `js/initcontext_test.go` line: 313 fails with this error:

```sh
--- FAIL: TestRequestWithBinaryFile (0.09s)
  Error Trace:    initcontext_test.go:361
  Error:          Received unexpected error:
                     GoError: post() can only be called from within default()
```

The js I'm testing:

```js
import http from "k6/http";
let binFile = open("/path/to/file.bin", "b");
export default function() {
  var data = {
    field: "this is a standard form field",
    file: http.file(binFile, "test.bin")
  };
  var res = http.post("%s", data);
  return true;
}
```

I tested the code outside of unit test for now at least. It is probably something simple - or not? :) - that elude me at the moment.

Couple of questions:

1. Are the `open()` function should always return []byte even for non binary file? Since the `crypto` and `encoding` function will need to be changed to accepting `[]byte` instead of `string`? Would that introduce breaking changes for user currently using `open()` and assuming they are receiving a string?

2. I think I'm missing why the filename should be optional when we use `http.file()`? The filename is how the part is identified when parsing the multipart message. For now I've set a naive "UnixName" if there's no filename specified. Maybe only the content-type should be optional?

Thanks for your help.

**Comment by liclac** [comment]
This should probably wait for #370 to be merged.

**Comment by afroozeh** [comment]
Is there any estimate when this PR will be merged? We're evaluating some load-testing tools and K6 is very attractive because we can write tests in JS, but we really need file upload...

**Comment by robingustafsson** [comment]
@afroozeh Yes, we intend to include this PR and https://github.com/loadimpact/k6/pull/370 in the next release that is planned to happen before the end of this month (Feb).

**Comment by robingustafsson** [comment]
I looked the test failure in `TestRequestWithBinaryFile`. It's caused by the lack of a proper `Context`, we need to initialize a full context with state and runtime before running the script as it's expected by the `HTTP` methods on the Go side, hence why the bridge binding code fails with the "GoError: post() can only be called from within default()" error (it looks for the existence of an initialized context in the closure, and yes the error message should be improved to point out the problem more clearly).

To answer the other questions:
1. No, what is currently implemented is fine. Open as text (`string`) by default, as binary (`[]byte`) if "b" is specified. Makes `open()` backwards compatible. If we can improve the handling of binary data in the future through an `ArrayBuffer` implementation in JS great but we're not there yet. The changes you did to the crypto module I find works as expected.
2. The filename should be optional I think, but the form field name should be required. Right now we set those fields to the same thing. I would separate them, require the `name` and have `filename` as an optional with a timestamp or similar as the default.

@dstpierre I will close https://github.com/loadimpact/k6/pull/370 in favor of this PR, do you want to finish it or should I take it over?

**Comment by spicykoala** [comment]
I have a fork for sending raw binary data as the request body - It heavily borrows from this PR so I'd like to wait for this one to be merged and then create my own PR from there.

Anything I can do to help move this along?

**Comment by robingustafsson** [comment]
@spicykoala I'm going to close this PR and create a new one based on it as the author of this PR is no longer active. Will ping you in the new PR.

**Pull request closed without merging**